pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,805 @@
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		if !apierrors.IsNotFound(err) {
+			c.queue.AddRateLimited(key)
+			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *Controller) syncPacketCapture(pcName string) error {
+	cleanupStatus := func() {
+		c.lock.Lock()
+		delete(c.captures, pcName)
+		c.lock.Unlock()
+	}
+
+	pc, err := c.packetCaptureLister.Get(pcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketCapture(pcName)
+			cleanupStatus()
+		}
+		return err
+	}
+
+	if err := c.validatePacketCapture(&pc.Spec); err != nil {
+		klog.ErrorS(err, "Invalid PacketCapture", "name", pc.Name)
+		if err := c.updatePacketCaptureStatus(pc.Name, 0, "", err); err != nil {
+			klog.ErrorS(err, "Failed to update PacketCapture status", "name", pc.Name)
+		}
+		cleanupStatus()
+		return nil
+	}
+
+	// Capture will not occur on this Node if a corresponding Pod interface is not found.
+	device := c.getTargetCaptureDevice(pc)
+	if device == nil {
+		return nil
+	}
+
+	status, err := func() (packetCapturePhase, error) {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		status := c.captures[pcName]
+		klog.InfoS("Syncing PacketCapture", "packagecapture", pcName, "status", status)
+		if status == packetCapturePhaseUnknown || status == packetCapturePhasePending {
+			klog.InfoS("Pending PackageCapture", "name", pcName)
+			newStatus, err := func() (packetCapturePhase, error) {
+				if c.numRunningCaptures < maxConcurrentCaptures {
+					c.numRunningCaptures += 1
+					return packetCapturePhaseRunning, c.startPacketCapture(pcName)
+				}
+				klog.InfoS("PacketCapture running count reach limit, waiting in queue", "name", pc.Name)
+				return packetCapturePhasePending, fmt.Errorf("PacketCapture running count reach limit, waiting in queue")
+			}()
+			c.captures[pcName] = newStatus
+			return newStatus, err
+		}
```

Comment:
There are several issues in the function, and it seems more complicated than needed, here is a revised version:
```go
status, err := func() (packetCapturePhase, error) {
	c.lock.Lock()
	defer c.lock.Unlock()
	status := c.captures[pcName]
	klog.InfoS("Syncing PacketCapture", "name", pcName, "status", status)
	if status != packetCapturePhasePending {
		return status, nil
	}
	if c.numRunningCaptures >= maxConcurrentCaptures {
		err = fmt.Errorf("PacketCapture running count reach limit")
	} else if err = c.startPacketCapture(pcName); err != nil {
		status = packetCapturePhaseCompleted
	} else {
		status = packetCapturePhaseRunning
	}
	c.captures[pcName] = status
	return status, err
}()
```